File: src/network/client/ServerClient.php
<?php

namespace pocketcloud\cloud\network\client;

use Closure;
use pocketcloud\cloud\console\log\CloudLogger;
use pocketcloud\cloud\network\Network;
use pocketcloud\cloud\network\packet\ClientboundPacket;
use pocketcloud\cloud\network\packet\CloudPacket;
use pocketcloud\cloud\Server;
use pocketcloud\cloud\server\CloudServer;
use pocketcloud\cloud\util\net\Address;

final class ServerClient {

    private array $delayedPackets = [];

    public function __construct(private readonly Address $address) {}

    public function sendPacket(ClientboundPacket $packet): bool {
        if (!Network::getInstance()->sendPacket($packet, $this)) {
            CloudLogger::get()->warn("Failed to send packet §b{} §rto §b{}§r.", $packet->getName(), $this->address);
            return false;
        }
        return true;
    }

    /** @internal */
    public function unsetDelayedPacket(int $index): void {
        if (isset($this->delayedPackets[$index])) {
            unset($this->delayedPackets[$index]);
            $this->delayedPackets = array_values($this->delayedPackets);
        }
    }

    /**
     * @param CloudPacket $packet
     * @param int $ticks delay in ticks (20 = 1s)
     * @param Closure(ServerClient $client, CloudPacket $packet, bool $success): void|null $onSend
     * @return void
     */
    public function sendDelayedPacket(CloudPacket $packet, int $ticks, ?Closure $onSend = null): void {
        $this->delayedPackets[] = [$packet, $Server::getInstance()->tick + $ticks, $onSend];
    }

    public function getDelayedPackets(): array {
        return $this->delayedPackets;
    }

    public function getAddress(): Address {
        return $this->address;
    }

    public function hasServer(): bool {
        return $this->getServer() !== null;
    }

    public function getServer(): ?CloudServer {
        return ServerClientCache::getInstance()->getServer($this);
    }

    public function toString(): string {
        return (string) $this;
    }

    public function __toString(): string {
        return "ServerClient[address=" . $this->getAddress() . "]";
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/network/client/ServerClient.java
package de.pocketcloud.cloud.network.client;

import de.pocketcloud.api.network.client.IServerClient;
import de.pocketcloud.api.network.packet.ClientboundPacket;
import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.server.CloudServer;
import io.netty.channel.Channel;
import io.netty.util.AttributeKey;
import lombok.Getter;
import lombok.experimental.Accessors;

import java.net.SocketAddress;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.TimeUnit;

public final class ServerClient implements IServerClient {

    public static final AttributeKey<ServerClient> ATTRIBUTE_KEY = AttributeKey.valueOf("serverClient");

    public record DelayedPacket(ClientboundPacket packet, long deliverAt, CompletableFuture<Void> future) {}

    @Getter
    @Accessors(fluent = true)
    private final Channel channel;
    private final Queue<DelayedPacket> delayedPackets = new ConcurrentLinkedQueue<>();

    public ServerClient(Channel channel) {
        this.channel = channel;
    }

    public CompletableFuture<Void> sendPacket(ClientboundPacket packet) {
        CompletableFuture<Void> future = new CompletableFuture<>();

        channel.writeAndFlush(packet).addListener(f -> {
            if (f.isSuccess()) {
                future.complete(null);
            } else {
                future.completeExceptionally(f.cause());
            }
        });

        return future;
    }

    @Override
    public CompletableFuture<Void> sendDelayedPacket(ClientboundPacket packet, long delay, TimeUnit unit) {
        CompletableFuture<Void> future = new CompletableFuture<>();
        delayedPackets.add(new DelayedPacket(packet, System.currentTimeMillis() + unit.toMillis(delay), future));
        return future;
    }

    @Override
    public CompletableFuture<Void> sendDelayedPacket(ClientboundPacket packet, long ticks) {
        CompletableFuture<Void> future = new CompletableFuture<>();
        delayedPackets.add(new DelayedPacket(packet, System.currentTimeMillis() + (ticks * 50), future));
        return future;
    }

    public List<DelayedPacket> pollDuePackets() {
        long now = System.currentTimeMillis();
        List<DelayedPacket> due = new ArrayList<>();
        delayedPackets.removeIf(dp -> {
            if (dp.deliverAt() <= now) {
                due.add(dp);
                return true;
            }
            return false;
        });

        return due;
    }

    public List<DelayedPacket> delayedPackets() {
        return List.copyOf(delayedPackets);
    }

    public SocketAddress address() {
        return channel.remoteAddress();
    }

    public boolean hasServer() {
        return server() != null;
    }

    public CloudServer server() {
        return (CloudServer) PocketCloud.instance().clients().getServer(this).orElse(null);
    }

    @Override
    public String toString() {
        return "ServerClient[address=" + address() + "]";
    }
}